gibberish_server/controllers/room_controller.js

```diff
@@ -24,36 +26,48 @@ class RoomController {
 
     static joinRoom(req, res) {
         // TODO: race conditionSTATE.GAME_WAITING
-        let r = roomModel.Room.deserializeRoom(roomModel.getRoomInfoAsJson(roomId));
-        if (r.state != STATE.GAME_WAITING) return "";
-        for (let p in r.players) {
-            if (p.name == nickname) {
-                return "";
-            }
-        }
-        r.players.push(new Player(nickname))
-        roomModel.saveRoom(r, client);
-        res.sen(r.players[r.length - 1].name);
+        const nickname = req.body.nickname
+        const roomId = req.body.roomId
+        roomModel.getRoomInfoAsJson(roomId)
+            .then(result => {
+                let r = roomModel.Room.deserializeRoom(result)
+                if (r.state != roomModel.STATE.GAME_WAITING) return "";
+                for (let p in r.players) {
+                    if (p.name == nickname) {
+                        return "";
+                    }
+                }
+                r.addPlayer(nickname)
+                roomModel.saveRoom(r);
+                res.send(r);
+            })
     }
 
     static startGame(req, res) {
         let roomId = req.body.roomId;
-        let r = roomModel.getRoomInfoAsObject(roomId);
-        r.start();
-        res.send(roomModel.saveRoom(r));
+        roomModel.getRoomInfoAsObject(roomId)
+            .then(room => {
+                room.start()
+                roomModel.saveRoom(room)
+                res.send(room)
+            })
     }
 
     static submitAnswer(req, res) {
         let roomId = req.body.roomId;
         let nickname = req.body.nickname;
-        let r = roomModel.getRoomInfoAsObject(roomId);
-        let score = new Date() - r.startedTime;
-        for (let p in r.playes) {
-            if (p.name == nickname) {
-                p.updateScore(score);
-                return
-            }
-        }
+        let score = req.body.score;
+        console.log(nickname)
```

Comment:
Good to add context with the logging, so that you know what it is when reading the log. e.g. console.log("nickname of current user", nickname)